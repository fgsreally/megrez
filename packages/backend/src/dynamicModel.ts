import { getModelForClass, prop } from '@typegoose/typegoose'
import { ProjectModel } from './dto/project.model'

export function createDynamicModel({ fields }: { fields: Record<string, any> }) {
  class DynamicModel {
    constructor(fields: Record<string, any>) {
      Object.assign(this, fields)
    }
  }

  for (const [fieldName, fieldOptions] of Object.entries(fields))

    prop(Object.assign(fieldOptions, { type: String }))(DynamicModel.prototype, fieldName)

  prop({ ref: () => ProjectModel })(DynamicModel.prototype, 'project')

  return getModelForClass(DynamicModel, { })
}

// // 示例：创建一个包含"name"和"age"字段的model
// const DynamicModel = createDynamicModel('myCollection', { name: { type: String }, age: { type: Number } })

// // 创建一个实例并保存到数据库
// const doc = new DynamicModel({ name: 'John', age: 30 })
// await doc.save()

// // 查询数据库
// const result = await DynamicModel.findOne({ name: 'John' })
