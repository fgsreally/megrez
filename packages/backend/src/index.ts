import { Factory, addGuard, bindApp } from 'phecda-server'
import express from 'express'
import { ConfigModule } from './modules/config/config.module'
import { AssetController } from './modules/asset/asset.controller'
import { NamespaceController } from './modules/namespace/namespace.controller'
import { TeamController } from './modules/team/team.controller'
import { jwtGuard } from './guards/jwt'
import { UserController } from './modules/user/user.controller'
const data = await Factory([ConfigModule, UserController, AssetController, NamespaceController, TeamController])
data.output()
const app = express()
// ServerContext.middlewareRecord.upload = uploadMiddleware
app.use(express.json())

addGuard('jwt', jwtGuard(data.moduleMap.get('user')!))

bindApp(app, data, { globalGuards: ['jwt'] })

export const viteNodeApp = app
