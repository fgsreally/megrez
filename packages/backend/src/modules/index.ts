export * from './user/user.controller'
export * from './user/user.model'
export * from './user/user.service'

export * from './team/team.controller'
export * from './team/team.model'
export * from './team/team.service'

export * from './namespace/namespace.controller'
export * from './namespace/namespace.model'
export * from './namespace/namespace.service'

export * from './config/config.module'

export * from './base/base.module'
export * from './base/base.controller'

export * from './asset/asset.controller'
export * from './asset/asset.model'
export * from './asset/asset.service'

export * from './file/file.controller'

export * from './record/record.model'
export * from './record/record.service'
export * from './record/record.controller'

export * from './logger/logger.service'
export * from './k8s/k8s.service'
export * from './email/email.service'
