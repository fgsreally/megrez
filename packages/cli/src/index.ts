import { PassThrough } from 'stream'
import archiver from 'archiver'
import axios from 'axios'
export async function uploadFilesAsZip(sourceFiles: string[], targetFileName: string, uploadUrl: string): Promise<void> {
  const output =new PassThrough()
  const archive = archiver('zip', { zlib: { level: 9 } })

  archive.on('error', (err) => {
    throw err
  })

  archive.pipe(output)

  for (const file of sourceFiles)
    archive.file(file, { name: file })

  archive.finalize()

  const data = new FormData()
  data.append('file', output.read(), targetFileName)

  try {
    const response = await axios.post(uploadUrl, data, {
      headers: {
        'Content-Type': 'multipart/form-data',
      },
    })

    console.log('Upload successful!')
    console.log(response.data)
  }
  catch (error) {
    console.error('Upload failed:', error)
  }
}

// 调用示例
// const sourceFiles = ['/path/to/file1', '/path/to/file2', '/path/to/file3'] // 指定要压缩的文件数组
// const targetFileName = 'output.zip' // 指定目标压缩文件名
// const uploadUrl = 'http://your-express-route-url' // 指定Express路由的URL

// uploadFilesAsZip(sourceFiles, targetFileName, uploadUrl)
//   .then(() => console.log('Done'))
//   .catch(err => console.error(err))
