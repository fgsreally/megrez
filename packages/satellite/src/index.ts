import { fork } from 'child_process'
import type { Socket } from 'net'
import { connect } from 'net'
import { resolve } from 'path'

import HP from 'http-proxy'
import { COMMAND } from './common'
import { sendMessage } from './utils'
const proxy = HP.createProxyServer({})
const pipeFile
  = process.platform === 'win32' ? '\\\\.\\pipe\\mypip' : '/tmp/unix.sock'

let client: Socket

function getPortFromBridge(id: string) {
  return sendMessage(client, {
    type: COMMAND.DEBUG,
    id,
  })
}

export function init() {
  const child = fork(resolve(__dirname, 'bridge.js'), {
    env: {
      MEGREZ_SATELLITE: 'true',
      ...process.env,
    },
  })
  child.stdout?.pipe(process.stdout)
  child.stderr?.pipe(process.stderr)

  child.once('message', () => {
    client = connect(pipeFile)
  })
}

export function debug(cmd: string, options?: {
  prefix?: string
  max?: number
  duration?: number
}) {
  const { max = 10, duration = 1000000, prefix = 'debug' } = options || {}

  const RE = new RegExp(`\/${prefix}\/(.+)\/(.*)`)

  return async (req: any, res: any) => {
    if (process.env.MEGREZ_SATELLITE)
      return
    const [, id, pathname] = req.originalUrl.match(RE) || []

    if (id) {
      try {
        const ret = await getPortFromBridge(id)

        req.url = `/${pathname}`
        return proxy.web(req, res, {
          target: `${req.protocol}://0.0.0.0:${ret!.port}`,
        })
      }
      catch (e) {
        res.status(404).send('not id')
      }
    }
    else {
      const { id: uid, path } = await sendMessage(client, {
        type: COMMAND.DEBUG,
        max,
        duration,
        cmd,
      })
      res.send(`
      <body>
      <script>
      const origin=location.origin.split('://')[1] 
      const debugEl=document.createElement('div')
      const baseEl =document.createElement('div')
     
      baseEl.innerHTML="baseUrl: "+origin+"/${prefix}/${uid}/"
      debugEl.innerHTML="debugUrl: devtools://devtools/bundled/js_app.html?experiments=true&v8only=true&ws="+origin+"/${prefix}/${uid}/${path}"
      document.body.append(baseEl)
      document.body.append(debugEl)
      </script>
      </body>
             `)
    }
  }
}

export function upgrade(options?: {
  prefix?: string
}) {
  const { prefix = 'debug' } = options || {}

  const RE = new RegExp(`\/${prefix}\/(.+)\/(.*)`)
  return async (req: any, socket: any, head: any) => {
    const [, id, pathname] = req.url.match(RE) || []
    if (id) {
      const ret = await getPortFromBridge(id)
      req.url = `/${pathname}`

      return proxy.ws(req, socket, head, {
        target: `ws://127.0.0.1:${ret!.debugPort}`,
      })
    }
  }
}

export function getToken(key: string, timeout: number) {
  return sendMessage(client, { key, timeout })
}
