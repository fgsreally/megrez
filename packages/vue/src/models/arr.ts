import { markRaw } from 'vue'
import { cloneDeep, isEqual, merge } from 'lodash-es'
export abstract class ArrModel<T extends object> {
  value: T[] = []
  _ADD = markRaw(new Set<T>())
  _REMOVE = markRaw(new Set<T>())
  _UPDATE = markRaw(new Set<T>())
  // update history
  _U_M = markRaw(new WeakMap<T, T>())
  unshift(item: T) {
    this.value.unshift(item)
    this._ADD.add(item)
  }

  push(item: T) {
    this.value.push(item)
    this._ADD.add(item)
  }

  pop() {
    const item = this.value.pop()
    this._REMOVE.add(item!)
    return item
  }

  _add(item: T) {
    this.value.unshift(item)
  }

  remove(item: T) {
    this._remove(item)
    this._REMOVE.add(item)
  }

  protected _remove(item: T) {
    const index = this.value.findIndex(i => item === i)
    if (!index)
      return
    this.value.splice(index, 1)
  }

  assign(oldItem: T, newItem: Partial<T>) {
    this._UPDATE.add(Object.assign(oldItem, newItem))
  }

  has(item: T) {
    const index = this.value.findIndex(i => item === i)
    return index >= 0
  }

  merge(item: T, part: Partial<T>) {
    if (this.has(item)) {
      this._UPDATE.add(item)
      this._U_M.set(item, cloneDeep(item))

      merge(item, part)
    }
  }

  updateOne(oldItem: T, newItem: T) {
    if (this._updateOne(oldItem, newItem)) {
      this._UPDATE.add(newItem)
      let history = this._U_H.get(oldItem)
      if (!history) {
        history = [cloneDeep(oldItem), cloneDeep(newItem)]
        this._U_H.set(oldItem, history!)
      }
      else {
        this._U_H.set(oldItem, [history[0], cloneDeep(newItem)])
      }
    }
  }

  protected _updateOne(oldItem: T, newItem: T) {
    const index = this.value.findIndex(item => item === oldItem)
    if (!index)
      return
    this.value.splice(index, 1, newItem)
    return true
  }

  exec() {
    for (const item of [...this._ADD]) {
      if (!this._REMOVE.has(item) && !this._UPDATE.has(item)) {
        this.create(item).catch(() => {
          this._add(item)
        }).finally(() => {
          this._REMOVE.delete(item)
        })
      }
    }
    for (const item of [...this._UPDATE]) {
      if (!this._REMOVE.has(item)) {
        const cloneItem = cloneDeep(item)
        this.update(item).then(() => {
          this._U_M.set(item, cloneItem)
          if (isEqual(item, cloneDeep))
            this._UPDATE.delete(item)
        }).catch(() => {
          if (isEqual(item, cloneDeep)) {
            this._updateOne(item, this._U_M.get(item)!)

            this._UPDATE.delete(item)
          }
        }).finally(() => {

        })
      }
    }

    for (const item of [...this._REMOVE]) {
      this.delete(item).then(() => {
        this._REMOVE.delete(item)
        this._ADD.delete(item)
        this._UPDATE.delete(item)
      })
    }
  }

  abstract create(item: T): Promise<void>
  abstract delete(item: T): Promise<void>
  abstract update(item: T): Promise<void>
  abstract find: (item: T) => Promise<void>
}
