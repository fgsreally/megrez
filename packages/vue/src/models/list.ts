import { markRaw } from 'vue'
import { cloneDeep, isEqual, merge } from 'lodash-es'

export class ListModel<T extends object> {
  value: T[] = []
  _ALL = markRaw(new WeakSet<T>())
  _ADD = markRaw(new Set<T>())
  _REMOVE = markRaw(new Set<T>())
  _UPDATE = markRaw(new Set<T>())
  // syncing item
  _S = markRaw(new WeakSet<T>())
  // update map
  _U_M = markRaw(new WeakMap<T, T>())

  isLocked: false | Promise<any> = false

  isSyncing(item: T) {
    if (this._S.has(item))
      return true
  }

  clean() {
    this._REMOVE.clear()
    this._UPDATE.clear()
    this._ADD.clear()
  }

  protected validate(_item: T) {
    return true
  }

  add(item: T) {
    if (this.has(item))
      return
    if (!this.validate(item))

      return

    this.value.unshift(item)
    this._ADD.add(item)
    this._ALL.add(item)
  }

  // push(item: T) {
  //   if (this.has(item))
  //     return
  //   if (!this.validate(item))

  //     return
  //   this.value.push(item)
  //   this._ADD.add(item)
  // }

  protected _add(item: T) {
    this._ALL.add(item)
    this.value.unshift(item)
  }

  remove(item: T) {
    if (!this.validate(item))

      return
    this._remove(item)
    this._REMOVE.add(item)
  }

  protected _remove(item: T) {
    const index = this.value.findIndex(i => item === i)
    if (index < 0)
      return
    this._ALL.delete(item)

    this.value.splice(index, 1)
  }

  protected _replace(oldItem: T, newItem: T) {
    const index = this.value.findIndex(item => item === oldItem)
    if (!index)
      return
    this._ALL.delete(oldItem)
    this._ALL.add(newItem)

    this.value.splice(index, 1, newItem)
    return true
  }

  has(item: T) {
    return this._ALL.has(item)
  }
}

export abstract class SyncListModel<T extends object> extends ListModel<T> {
  merge(item: T, part: Partial<T>) {
    if (!this.has(item))
      return

    if (!this.validate(item))
      return
    this._UPDATE.add(item)
    this._U_M.set(item, cloneDeep(item))

    merge(item, part)
  }

  protected validate(_item: T): boolean {
    return !this.isSyncing(_item) && !this.isLocked
  }

  async push() {
    if (this.isLocked)
      await this.isLocked
    return this.isLocked = new Promise<void>((resolve) => {
      let count = 0
      for (const item of [...this._UPDATE]) {
        if (!this._REMOVE.has(item) && !this._ADD.has(item)) {
          this._S.add(item)
          count++
          this.update(item).catch(() => {
            this._replace(item, this._U_M.get(item)!)
          }).finally(() => {
            if (--count === 0)
              resolve()
            this._S.delete(item)
            this._UPDATE.delete(item)
          })
        }
      }

      for (const item of [...this._ADD]) {
        if (!this._REMOVE.has(item) && !this._S.has(item)) {
          this._S.add(item)
          count++

          this.create(item).catch(() => {
            this._remove(item)
          }).finally(() => {
            this._ADD.delete(item)
            this._S.delete(item)
            if (--count === 0)
              resolve()

            this._REMOVE.delete(item)
          })
        }
      }

      for (const item of [...this._REMOVE]) {
        this._S.add(item)
        count++
        this.delete(item).catch(() => {
          this._add(item)
        }).finally(() => {
          if (--count === 0)
            resolve()
          this._REMOVE.delete(item)
          this._ADD.delete(item)
          this._UPDATE.delete(item)
        })
      }
    }).then(() => {
      this.isLocked = false
    })
  }

  async pull() {
    await this.isLocked
    this.isLocked = this.find()
    this.value = await this.isLocked

    this.isLocked = false
  }

  abstract create(item: T): Promise<void>
  abstract delete(item: T): Promise<void>
  abstract update(item: T): Promise<void>
  abstract find(): Promise<T[]>
}

export abstract class AsyncListModel<T extends object> extends ListModel<T> {
  merge(item: T, part: Partial<T>) {
    if (!this.has(item))
      return

    if (!this.validate(item))
      return
    this._UPDATE.add(item)
    this._U_M.set(item, cloneDeep(item))

    merge(item, part)
  }

  async push() {
    if (this.isLocked)
      await this.isLocked
    return this.isLocked = new Promise<void>((resolve) => {
      let count = 0
      const update = (item: T) => {
        if (this._S.has(item))
          return
        this._S.add(item)

        if (this._REMOVE.has(item)) {
          count++
          this.delete(item).catch(() => {
            this._add(item)
          }).finally(() => {
            if (--count === 0)
              resolve()
            this._S.delete(item)
            this._REMOVE.delete(item)
            this._ADD.delete(item)
            this._UPDATE.delete(item)
          })
          return
        }

        if (this._ADD.has(item)) {
          count++
          this._UPDATE.delete(item)
          this.create(item).catch(() => {
            this._remove(item)
          }).finally(() => {
            this._S.delete(item)
            this._ADD.delete(item)
            update(item)
            if (--count === 0)
              resolve()
          })
          return
        }
        if (this._UPDATE.has(item)) {
          count++

          this._S.add(item)
          const cloneItem = cloneDeep(item)
          this.update(item).then(() => {
            this._S.delete(item)

            this._U_M.set(item, cloneItem)
            if (isEqual(item, cloneItem))
              this._UPDATE.delete(item)
            else update(item)
            if (--count === 0)
              resolve()
          }).catch(() => {
            this._S.delete(item)

            if (isEqual(item, cloneDeep)) {
              this._replace(item, this._U_M.get(item)!)

              this._UPDATE.delete(item)
            }
            else { update(item) }
            if (--count === 0)
              resolve()
          })
        }
      }
      const items = [...this._UPDATE, ...this._ADD, ...this._REMOVE]

      if (!items.length)
        resolve()
      for (const item of items) update(item)
    }).then(() => {
      this.isLocked = false
    })
  }

  async pull() {
    await this.isLocked
    this.isLocked = this.find()
    this.value = await this.isLocked

    this.isLocked = false
  }

  abstract create(item: T): Promise<void>
  abstract delete(item: T): Promise<void>
  abstract update(item: T): Promise<void>
  abstract find(): Promise<T[]>
}
