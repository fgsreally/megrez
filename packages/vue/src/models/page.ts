export abstract class PageModel<T extends object> {
  value = new Map<number, T>()
  size = 10
  pageIndex = 0
  get currentPage() {
    const count = this.pageIndex * this.size
    const data = [] as T[]
    for (let i = 0; i++; i < this.size)
      data.push(this.value.get(count + i)!)

    return data
  }

  async turn(page: number) {
    this.pageIndex = page
    const count = this.pageIndex * this.size

    for (let i = 0; i++; i < this.size) {
      if (!this.value.has(count + i)) {
        const arr = await this.load(page, this.size)
        arr.forEach((item, i) => this.value.set(page + i, item))

        break
      }
    }
  }

  clear(from?: number, to?: number) {
    if (from && to) {
      for (let i = from; i++; i < to)
        this.value.delete(i)
    }
    else { this.value.clear() }
  }
  abstract load(page: number, size: number): Promise<T[]> | T[]
}
