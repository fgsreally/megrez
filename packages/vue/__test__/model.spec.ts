import { describe, expect, it, vi } from 'vitest'
import { cloneDeep, isEqual } from 'lodash-es'
import { AsyncListModel, SyncListModel } from '../src/models'
describe('model', () => {
  it('sync list model', async () => {
    const errorFn = vi.fn()
    const createFn = vi.fn()
    const updateFn = vi.fn()
    const deleteFn = vi.fn()

    const data = [{
      id: 1,
    }, { id: 2 }]

    interface Item { id: number; value?: string }
    class TestModel extends SyncListModel<Item> {
      async find() {
        return cloneDeep(data)
      }

      protected validate(_item: Item): boolean {
        const ret = super.validate(_item)
        if (!ret)
          errorFn()
        return ret
      }

      async create(item: Item) {
        createFn()
        data.unshift(item)
      }

      async delete(item: Item) {
        deleteFn()
        const index = data.findIndex(({ id }) => item.id === id)
        index > -1 && data.splice(index, 1)
      }

      async update(item: Item) {
        updateFn()
        const index = data.findIndex(({ id }) => item.id === id)
        index > -1 && data.splice(index, 1, item)
      }
    }
    const instance = new TestModel()
    const item1 = {
      id: 3,
      value: 'item1',
    }
    const item2 = {
      id: 4,
      value: 'item2',
    }
    // pull
    await instance.pull()
    expect(instance.value).toMatchSnapshot()
    // create>update
    instance.add(item1)
    instance.merge(item1, { value: 'new item1' })
    instance.push()
    await instance.isLocked
    expect(createFn).toBeCalledTimes(1)
    expect(updateFn).toBeCalledTimes(0)
    expect(data).toMatchSnapshot()
    // delete>create
    instance.add(item2)
    expect(isEqual(instance.value, data)).toBe(false)

    instance.remove(item2)
    instance.push()
    await instance.isLocked
    expect(createFn).toBeCalledTimes(1)
    expect(deleteFn).toBeCalledTimes(1)
    expect(data).toMatchSnapshot()
    expect(isEqual(instance.value, data)).toBe(true)

    // won't act item or pull which during push
    expect(errorFn).toBeCalledTimes(0)
    instance.add(item2)
    instance.push()
    instance.pull()
    instance.delete(item2)
    expect(errorFn).toBeCalledTimes(1)
  })
  it('async list model', async () => {
    const createFn = vi.fn()
    const updateFn = vi.fn()
    const deleteFn = vi.fn()

    const data = [{
      id: 1,
    }, { id: 2 }]

    interface Item { id: number; value?: string }
    class TestModel extends AsyncListModel<Item> {
      async find() {
        return cloneDeep(data)
      }

      async create(item: Item) {
        createFn()
        data.unshift(item)
      }

      async delete(item: Item) {
        deleteFn()
        const index = data.findIndex(({ id }) => item.id === id)
        index > -1 && data.splice(index, 1)
      }

      async update(item: Item) {
        const index = data.findIndex(({ id }) => item.id === id)
        if (index > -1) {
          data.splice(index, 1, item)

          updateFn()
        }
      }
    }
    const instance = new TestModel()
    const item1 = {
      id: 3,
      value: 'item1',
    }
    const item2 = {
      id: 4,
      value: 'item2',
    }
    const item3 = {
      id: 5,
      value: 'item3',
    }
    // pull
    await instance.pull()
    expect(instance.value).toMatchSnapshot()
    // create>update
    instance.add(item1)
    instance.merge(item1, { value: 'new item1' })
    await instance.push()
    expect(createFn).toBeCalledTimes(1)
    expect(updateFn).toBeCalledTimes(0)

    expect(data).toMatchSnapshot()
    // delete>create
    instance.add(item2)
    expect(isEqual(instance.value, data)).toBe(false)

    instance.remove(item2)
    await instance.push()
    expect(deleteFn).toBeCalledTimes(1)
    expect(data).toMatchSnapshot()
    expect(isEqual(instance.value, data)).toBe(true)

    instance.add(item3)
    instance.push()
    instance.merge(item3, { value: 'new item3' })
    expect(createFn).toBeCalledTimes(2)

    await instance.push()

    expect(updateFn).toBeCalledTimes(1)

    expect(isEqual(instance.value, data)).toBe(true)
    data.shift({
      id: '6',
      value: 'item6',
    })
    expect(isEqual(instance.value, data)).toBe(false)

    await instance.pull()
    expect(isEqual(instance.value, data)).toBe(true)
  })
})
