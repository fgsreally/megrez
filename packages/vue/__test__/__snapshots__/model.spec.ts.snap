// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`model > async list model 1`] = `
[
  {
    "id": 1,
  },
  {
    "id": 2,
  },
]
`;

exports[`model > async list model 2`] = `
[
  {
    "id": 3,
    "value": "new item1",
  },
  {
    "id": 1,
  },
  {
    "id": 2,
  },
]
`;

exports[`model > async list model 3`] = `
[
  {
    "id": 3,
    "value": "new item1",
  },
  {
    "id": 1,
  },
  {
    "id": 2,
  },
]
`;

exports[`model > sync list model 1`] = `
[
  {
    "id": 1,
  },
  {
    "id": 2,
  },
]
`;

exports[`model > sync list model 2`] = `
[
  {
    "id": 3,
    "value": "new item1",
  },
  {
    "id": 1,
  },
  {
    "id": 2,
  },
]
`;

exports[`model > sync list model 3`] = `
[
  {
    "id": 3,
    "value": "new item1",
  },
  {
    "id": 1,
  },
  {
    "id": 2,
  },
]
`;
